### ISSUE TITLE
vkWaitForFences needs to be specified as a thread cancellation point

### ISSUE THREAD

**Issue by msharov** [KhronosGroup/Vulkan-Docs#2107]
When a vkQueueSubmit signal fence is waited on by vkWaitForFences on a separate thread, as is always required in any application that needs to wait for user input, cancelling the thread results in validation error when the fence is destroyed:
```
validation layer: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0x4a581c0000000195, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x141cb623 | THREADING ERROR : vkDestroyFence(): object of type VkFence is simultaneously used in thread 140624706307776 and thread 140624945473344
validation layer: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0x55c492d31c00, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x141cb623 | THREADING ERROR : vkDestroyDevice(): object of type VkDevice is simultaneously used in thread 140624706307776 and thread 140624945473344
validation layer: Unloading layer library libVkLayer_khronos_validation.so
```
As a workaround, it is possible to make the thread uncancellable at that point:
```c++
pthread_setcancelstate (PTHREAD_CANCEL_DISABLE, nullptr);
vkWaitForFences (_device, _fences.size(), _fences.begin(), VK_FALSE, 100000000);
pthread_setcancelstate (PTHREAD_CANCEL_ENABLE, nullptr);
```
which eliminates the error by forcing the wait to complete. This indicates that `vkWaitForFences` implementation does not handle cancellation properly.

The workaround of disabling cancellation is not optimal because if `vkWaitForFences` hangs for whatever reason, then so will pthread_join, and so will the process trying to clean it up.

Furthermore, this gotcha is not mentioned in `vkWaitForFences` documentation, and as it appears to most of us to be analogous to `poll`, which is a cancellation point, that is how we would try to use it.

I request amending the specification for `vkWaitForFences` to be explicitly required to be cancellable.

**Comment by r-potter** [comment]
At minimum this would require an extension, because it's a functional change that's unlikely to get backported onto all drivers.

More generally, this seems like something that will break layers that expect a PreHook -> underlying vkWaitForFences -> PostHook execution sequence and don't get a chance to clean up when threads get cancelled, This appears to be exactly what you are describing in the first issue. The validation layer added some tracking information in the PreHook, didn't get a chance to delete in the PostHook, and so the layer now thinks there is a dangling reference.

It's hard for me to see how layer authors in general can reasonably be expected to handle thread cancellation.

**Comment by cubanismo** [comment]
I assume this is essentially a request that grew out of discussion and suggestions on https://github.com/KhronosGroup/Vulkan-Docs/issues/310, so it's probably worth weighing the difficulty of defining something to resolve this issue, Vs. prioritizing that issue.

**Comment by msharov** [comment]
Not requiring cancellation imposes a lot of complexity on the application, from the API user's point of view. Consider all the various vkWait functions; vkWaitForFences, vkWaitSemaphores, and vkWaitForPresentKHR may all be needed in the render loop. Then there is the main event loop fd poll, and whatever thread synchronization primitives you want to use.

Because all these wait for different things, each one can only wait for one kind of thing. Because the application always needs to wait for more than one kind of thing, vkWaitForFences/Present and fd poll at the very minimum, it needs to have a separate thread for each kind of wait. Each of these threads would be stuck waiting for its thing, being unable to do anything else, importantly including being able to check if waiting for new things is requested by other threads, unless you timeout regularly and poll, in which case you halfway back to busywait polling.

Each thread will need to use a communication framework to get the wait objects submitted and then returned. The easiest is to create a socketpair for each thread and send the wait objects over it, first to the thread to be waited on, then returned when the wait is completed. Without thread cancellation, you must also implement a "quit" command sent alongside, such as sending a null VkFence. The socket must be nonblocking, because you also want to wait vkWaitForFences. If you want to be able to add new fences or quit at an arbitrary point, vkWaitForFences must also timeout at reasonable intervals and poll its socket for new fences.

Async socket reads/writes are a lot of code; you need to handle partial transfers, so you'll need to maintain four buffers, two per socket end. You'll need to handle these partial reads and writes, poll the fd for both reading and writing, handle EAGAIN, EINTR, ECONNRESET, and think of some sane way to report or handle other errors. Because the threads must be persistent, you'll more code to spawn them on submit and restart them on error (or your app will hang thinking it's waiting on a fence).

All this complexity and the inevitable problems of multithreading can make anyone lose his hair. Do you really still think that having so many different kinds of waits is so much simpler than just using fds for everything?

**Comment by TomOlson** [comment]
The Working Group discussed in a weekly call on 2023-04-19. Strong consensus from driver and layer writers that they can't see a remotely robust way to do this. We agree with @cubanismo's [comment](https://github.com/KhronosGroup/Vulkan-Docs/issues/2107#issuecomment-1513820680) above, that issue #310 (now seven years old) is the direction this would need to go. We are always interested in use cases as a basis for prioritizing spec development work - thanks @msharov for explaining yours, happy to know of others.

**Comment by cubanismo** [comment]
> Not requiring cancellation imposes additional complexity on the wait thread management code and exit cleanup. [...]

This is a bit dire sounding.

I think I may have inadvertently painted this as an either/or choice above. Note EGL and OpenGL have the same issues (encapsulated CPU waits via glFinish(), eglClientWaitSync(), etc.) with no provisions for cancellation, and no one's ever complained that I know of. In my view, Vulkan isn't any worse than a multitude of other APIs in this regard, so doing nothing would be a reasonable resolution as well. Applications would just be required to use reasonable timeouts in their waits and work with the fact that they can't cancel threads running Vulkan commands, which will still be the case with the proposed resolution. Generally, cancelling threads isn't a great idea if those threads do anything non-trivial. I don't know of any libraries off-hand that define safe cancellation points above the level of libc. POSIX gives applications tools to deal with this, in that they can guard unsafe code sections against cancellation (An app could disable cancellation around every individual Vulkan call, for example, though this would be pretty ridiculous. They'd more likely disable cancellation, run a sequence of Vulkan commands, then re-enable it). That doesn't help a thread sitting in vkWaitForFences() indefinitely, but you could very reasonably have a loop like this:

```C
while (1) {
   int oldCancelState, unused;
   pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldCancelState);
   vkWaitForFences(device, numFences, fences, VK_FALSE, TIMEOUT_100MS);
   pthread_setcancelstate(oldCancelState, &unused);
}
```

As your threadFunc.

That said, no opposition to moving forward with #310. It'd be a nice extension to have.

**Comment by msharov** [comment]
I have rewritten the comment several times to make it less dire, but my frustration can't help but to show through :) There is a philosophical problem here, as well as a practical one. One of the benefits frequently touted for the Vulkan API is that it can avoid various bottlenecks inevitable in the single-state OpenGL.

In OpenGL, nobody complains about the wait functions being non-cancellable because you can't use multithreading in OpenGL in the first place. You can't have a separate thread waiting for a query object indicating that the frame has finished rendering. What I did instead was to guess the likely frame duration and wait in poll until then. This requires tracking the refresh rate, and does not work at all if you have a variable refresh rate monitor.

Vulkan was supposed to get rid of things like that and allow both the CPU and the GPU to complete the maximum possible amount of work by doing it in parallel and by explicit synchronization only where required. Here, on the other hand, vkWaitForFences feels like a slap in the face, forcing the CPU to sit and wait for the frame to render, while it could be doing something useful instead. Yes, you can timeout or thread it, but that just makes it a busywait poll like the OpenGL query objects, which is what we came to Vulkan to avoid. Workarounds like this make Vulkan work, but psychologically they hurt.

**Comment by cubanismo** [comment]
> In OpenGL, nobody complains about the wait functions being non-cancellable because you can't use multithreading in OpenGL in the first place. You can't have a separate thread waiting for a query object indicating that the frame has finished rendering. What I did instead was to guess the likely frame duration and wait in poll until then. This requires tracking the refresh rate, and does not work at all if you have a variable refresh rate monitor.

That's not true. You can't use glFinish() or GL query objects, but you could use GL_ARB_sync with share lists or EGL_KHR_fence_sync/EGL 1.5 to wait for completion across threads in roughly the same manner as VkFence. GL has its shortcomings and notably a history of driver bugs with multithreading, but its not that functionally different in this particular area.

> Yes, you can timeout or thread it, but that just makes it a busywait poll like the OpenGL query objects, which is what we came to Vulkan to avoid. Workarounds like this make Vulkan work, but psychologically they hurt.

I understand it isn't as theoretically clean, but I would indeed expect it to work. A blocking call with a reasonable timeout will perform much better (performance and power-wise) than pure busy-waiting, and I think it would be disingenuous to equate the two outside the scope psychological pain. If we were talking about a use case where you needed very high responsiveness, I wouldn't suggest something with long timeouts (many ms) like this, but for a situation like thread teardown that is probably rare and/or an exception handling situation, and which I'm assuming has no constraints other than remaining "interactive" as perceived by a human, a ~10 or ~100ms delay probably wouldn't be an issue in practice. Of course, I don't know for sure if those assumptions are correct for your use case, or in other use cases.

Regardless, I sympathize, and as Tom noted above, I also do personally appreciate you pointing out this cancellation issue, as I (and I assume others) hadn't considered it previously. I made further comments simply because I think they'd useful when prioritizing work. Having a "clean" spec should also carry significant weight in prioritization exercises though. IMHO, considering known use cases helps most of the time with today's usage, but spec/API cleanliness helps the use cases no one (or at least, no one in the Vulkan working group) has thought of yet, and I prefer my low level APIs to make a reasonable attempt to handle such situations. They wouldn't be much use otherwise. As has been made clear here (and countless times in the past), it's impossible to know all potential use cases/patterns.

**Comment by oddhack** [comment]
@TomOlson I've had this on the maintenance call agenda for a long time with a note that we were going to ask @cubanismo to suggest a path forward on this. Any further thoughts?

**Comment by versalinyaa** [comment]
Closing as WontFix, as explained in <https://github.com/KhronosGroup/Vulkan-Docs/issues/2107#issuecomment-1515518405> and <https://github.com/KhronosGroup/Vulkan-Docs/issues/2107#issuecomment-1514637035>.
